src/scripts/Deploy-Toolkit.ps1

```diff
@@ -1,53 +1,150 @@
+    Name of the template or module to deploy. Default = finops-hub.
 .PARAMETER ResourceGroup
     Optional. Name of the resource group to deploy to. Will be created if it doesn't exist. Default = ftk-<username>-<computername>.
 .PARAMETER Location
-    Optional. Azure location to deploy the resource group from. Not used for resources. Resources use the resource group location. Default = westus.
+    Optional. Azure location to execute the deployment from. Default = westus.
+.PARAMETER Parameters
+    Optional. Parameters to pass thru to the deployment.
+.PARAMETER Build
+    Optional. Indicates whether the the Build-Toolkit command should be executed first. Default = false.
+.PARAMETER Test
+    Optional. Indicates whether to run the template or module test instead of the template or module itself. Default = false.
 .PARAMETER WhatIf
     Optional. Displays a message that describes the effect of the command, instead of executing the command.
 #>
 Param(
     [Parameter(Position = 0)][string]$Template = "finops-hub",
     [string]$ResourceGroup,
     [string]$Location = "westus",
+    [object]$Parameters,
+    [switch]$Build,
+    [switch]$Test,
     [switch]$WhatIf
 )
 
-# Set default RG name to "ftk-<username>-<computername>"
-If ([string]::IsNullOrEmpty($ResourceGroup)) {
-    # For some reason, using variables directly does not get the value until we write them
+# Use the debug flag from common parameters to determine whether to run in debug mode
+$Debug = $DebugPreference -eq "Continue"
+
+function iff([bool]$Condition, $IfTrue, $IfFalse) {
+    if ($Condition) { $IfTrue } else { $IfFalse }
+}
+
+# Build toolkit if requested
+if ($Build) {
+    ./Build-Toolkit
+}
+
+# Generates a unique name based on the signed in username and computer name for local testing
+function Get-UniqueName() {
+    # NOTE: For some reason, using variables directly does not get the value until we write them
     $c = $env:ComputerName
     $u = $env:USERNAME
     $c | Out-Null
     $u | Out-Null
-    $ResourceGroup = "ftk-$u-$c".ToLower()
+    return "ftk-$u-$c".ToLower()
 }
 
-# Create resource group if it doesn't exist
-$rg = Get-AzResourceGroup $ResourceGroup
-If ($null -eq $rg) {
-    New-AzResourceGroup `
-        -Name $ResourceGroup `
-        -Location $Location `
-    | Out-Null
+# Local dev parameters
+$defaultParameters = @{
+    "finops-hub"      = @{ hubName = Get-UniqueName }
+    "finops-hub/test" = @{ uniqueName = Get-UniqueName }
 }
 
-# Start deployment
-$params = @{ hubName = $ResourceGroup };
-New-AzResourceGroupDeployment `
-    -TemplateFile "../templates/$Template/main.bicep" `
-    -TemplateParameterObject $params `
-    -ResourceGroupName $ResourceGroup `
-    -WhatIf:$WhatIf
+# Find bicep file (templates first)
+@("../templates", "../bicep-registry", "../../release") `
+| ForEach-Object { Get-Item (Join-Path $_ $Template (iff $Test test/main.test.bicep main.bicep)) -ErrorAction SilentlyContinue } `
+| ForEach-Object {
```

Comment:
Somewhere in this file is where we should be recording deployments via a logging API of some sort, can we add a placeholder?

Reply:
The Bicep Registry already has a mechanism for tracking deployments, so we shouldn't need one. That said, we do need telemetry outside of the Bicep Registry (like for the finops-hub template). I added an issue to track this: #89.